In [ ]:
!pip install -U scikit-learn
!pip install pandas

In [38]:
import pandas as pd

In [46]:
#  reading the spam_ham dataset.
df = pd.read_csv("dataset/spam_ham_dataset.csv")
df

Unnamed: 0 label                                               text  \
0            605   ham  Subject: enron methanol ; meter # : 988291\r\n...   
1           2349   ham  Subject: hpl nom for january 9 , 2001\r\n( see...   
2           3624   ham  Subject: neon retreat\r\nho ho ho , we ' re ar...   
3           4685  spam  Subject: photoshop , windows , office . cheap ...   
4           2030   ham  Subject: re : indian springs\r\nthis deal is t...   
...          ...   ...                                                ...   
5166        1518   ham  Subject: put the 10 on the ft\r\nthe transport...   
5167         404   ham  Subject: 3 / 4 / 2000 and following noms\r\nhp...   
5168        2933   ham  Subject: calpine daily gas nomination\r\n>\r\n...   
5169        1409   ham  Subject: industrial worksheets for august 2000...   
5170        4807  spam  Subject: important online banking alert\r\ndea...   

      label_num  
0             0  
1             0  
2             0  
3             1  
4             0  
...         ...  
5166          0  
5167          0  
5168          0  
5169          0  
5170          1  

[5171 rows x 4 columns]

In [47]:
# getting the columns of the dataset
df.columns

Index(['Unnamed: 0', 'label', 'text', 'label_num'], dtype='object')

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5171 entries, 0 to 5170
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  5171 non-null   int64 
 1   label       5171 non-null   object
 2   text        5171 non-null   object
 3   label_num   5171 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 161.7+ KB


In [49]:
# first column is useless so need to remove it.
del df['Unnamed: 0']

In [50]:
df.head()

label                                               text  label_num
0   ham  Subject: enron methanol ; meter # : 988291\r\n...          0
1   ham  Subject: hpl nom for january 9 , 2001\r\n( see...          0
2   ham  Subject: neon retreat\r\nho ho ho , we ' re ar...          0
3  spam  Subject: photoshop , windows , office . cheap ...          1
4   ham  Subject: re : indian springs\r\nthis deal is t...          0

In [51]:
# 'label' column contain two type of unique values
df['label'].unique()

array(['ham', 'spam'], dtype=object)

In [52]:
# 'text' column contains the texts 
df['text']

0       Subject: enron methanol ; meter # : 988291\r\n...
1       Subject: hpl nom for january 9 , 2001\r\n( see...
2       Subject: neon retreat\r\nho ho ho , we ' re ar...
3       Subject: photoshop , windows , office . cheap ...
4       Subject: re : indian springs\r\nthis deal is t...
                              ...                        
5166    Subject: put the 10 on the ft\r\nthe transport...
5167    Subject: 3 / 4 / 2000 and following noms\r\nhp...
5168    Subject: calpine daily gas nomination\r\n>\r\n...
5169    Subject: industrial worksheets for august 2000...
5170    Subject: important online banking alert\r\ndea...
Name: text, Length: 5171, dtype: object

In [53]:
# first text
df['text'][0]

"Subject: enron methanol ; meter # : 988291\r\nthis is a follow up to the note i gave you on monday , 4 / 3 / 00 { preliminary\r\nflow data provided by daren } .\r\nplease override pop ' s daily volume { presently zero } to reflect daily\r\nactivity you can obtain from gas control .\r\nthis change is needed asap for economics purposes ."

In [10]:
# contain the label into the numeric form
df['label_num']

0       0
1       0
2       0
3       1
4       0
       ..
5166    0
5167    0
5168    0
5169    0
5170    1
Name: label_num, Length: 5171, dtype: int64

In [54]:
# 0 in label_num means ham
df[df['label_num'] == 0]['label'].unique()

array(['ham'], dtype=object)

In [55]:
# 1 in label_num means spam
df[df['label_num'] == 1]['label'].unique()

array(['spam'], dtype=object)

In [56]:
df.head()

label                                               text  label_num
0   ham  Subject: enron methanol ; meter # : 988291\r\n...          0
1   ham  Subject: hpl nom for january 9 , 2001\r\n( see...          0
2   ham  Subject: neon retreat\r\nho ho ho , we ' re ar...          0
3  spam  Subject: photoshop , windows , office . cheap ...          1
4   ham  Subject: re : indian springs\r\nthis deal is t...          0

In [57]:
# total rows
df.shape[0]

5171

In [60]:
# deleting label columns no need of because it give same information as given by the label_num column
del df['label']

In [61]:
# checking the null value in each row
df.isnull().sum()

text         0
label_num    0
dtype: int64

In [62]:
'''
0 --> ham
1 --> spam
'''
df['label_num'].value_counts()

0    3672
1    1499
Name: label_num, dtype: int64

In [19]:
X = df["text"]
y = df['label_num']

In [63]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,
                                                    test_size=0.2,
                                                    random_state=42,
                                                    shuffle=True
                                                   )

In [64]:
print("Input train: ", X_train.shape)
print("Input test: ", X_test.shape)
print("Output train: ", y_train.shape)
print("Output test: ", y_test.shape)


Input train:  (4136,)
Input test:  (1035,)
Output train:  (4136,)
Output test:  (1035,)


In [65]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import make_pipeline
model = make_pipeline(TfidfVectorizer(),MultinomialNB())

In [66]:
model.fit(X_train,y_train)

Pipeline(steps=[('tfidfvectorizer', TfidfVectorizer()),
                ('multinomialnb', MultinomialNB())])

In [67]:
y_pred = model.predict(X_test)
y_pred

array([0, 1, 0, ..., 1, 0, 0])

In [68]:
from sklearn.metrics import accuracy_score
acc =  accuracy_score(y_pred, y_test)
acc

0.8772946859903382

In [73]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[742,   0],
       [127, 166]])

In [74]:
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
tn, fp, fn, tp

(742, 0, 127, 166)

- Total False Negative = 127. 
- FN means actual text is of spam but predicted ham because dataset is not balance. hams are greater than spams so model is overfitted toward the hams.